waku/waku_archive_legacy/archive.nim

```diff
@@ -0,0 +1,326 @@
+
+## Archive
+
+type WakuArchive* = ref object
+  driver: ArchiveDriver
+
+  validator: MessageValidator
+
+  retentionPolicy: Option[RetentionPolicy]
+
+  retentionPolicyHandle: Future[void]
+  metricsHandle: Future[void]
+
+proc validate*(msg: WakuMessage): Result[void, string] =
+  if msg.ephemeral:
+    # Ephemeral message, do not store
+    return
+
+  if msg.timestamp == 0:
+    return ok()
+
+  let
+    now = getNanosecondTime(getTime().toUnixFloat())
+    lowerBound = now - MaxMessageTimestampVariance
+    upperBound = now + MaxMessageTimestampVariance
+
+  if msg.timestamp < lowerBound:
+    return err(invalidMessageOld)
+
+  if upperBound < msg.timestamp:
+    return err(invalidMessageFuture)
+
+  return ok()
+
+proc new*(
+    T: type WakuArchive,
+    driver: ArchiveDriver,
+    validator: MessageValidator = validate,
+    retentionPolicy = none(RetentionPolicy),
+): Result[T, string] =
+  if driver.isNil():
+    return err("archive driver is Nil")
+
+  let archive =
+    WakuArchive(driver: driver, validator: validator, retentionPolicy: retentionPolicy)
+
+  return ok(archive)
+
+proc handleMessage*(
+    self: WakuArchive, pubsubTopic: PubsubTopic, msg: WakuMessage
+) {.async.} =
+  self.validator(msg).isOkOr:
+    waku_legacy_archive_errors.inc(labelValues = [error])
+    return
+
+  let
+    msgDigest = computeDigest(msg)
+    msgDigestHex = msgDigest.data.to0xHex()
+    msgHash = computeMessageHash(pubsubTopic, msg)
+    msgHashHex = msgHash.to0xHex()
+    msgTimestamp =
+      if msg.timestamp > 0:
+        msg.timestamp
+      else:
+        getNanosecondTime(getTime().toUnixFloat())
+
+  trace "handling message",
+    msg_hash = msgHashHex,
+    pubsubTopic = pubsubTopic,
+    contentTopic = msg.contentTopic,
+    msgTimestamp = msg.timestamp,
+    usedTimestamp = msgTimestamp,
+    digest = msgDigestHex
+
+  let insertStartTime = getTime().toUnixFloat()
+
+  (await self.driver.put(pubsubTopic, msg, msgDigest, msgHash, msgTimestamp)).isOkOr:
+    waku_legacy_archive_errors.inc(labelValues = [insertFailure])
+    error "failed to insert message", error = error
+
+  debug "message archived",
+    msg_hash = msgHashHex,
+    pubsubTopic = pubsubTopic,
+    contentTopic = msg.contentTopic,
+    msgTimestamp = msg.timestamp,
+    usedTimestamp = msgTimestamp,
+    digest = msgDigestHex
+
+  let insertDuration = getTime().toUnixFloat() - insertStartTime
+  waku_legacy_archive_insert_duration_seconds.observe(insertDuration)
+
+proc findMessages*(
+    self: WakuArchive, query: ArchiveQuery
+): Future[ArchiveResult] {.async, gcsafe.} =
+  ## Search the archive to return a single page of messages matching the query criteria
+
+  let maxPageSize =
+    if query.pageSize <= 0:
+      DefaultPageSize
+    else:
+      min(query.pageSize, MaxPageSize)
+
+  let isAscendingOrder = query.direction.into()
+
+  if query.contentTopics.len > 10:
+    return err(ArchiveError.invalidQuery("too many content topics"))
```

Comment:
It is arbitrary! I will change that in the new version.

Reply:
> It is arbitrary! I will change that in the new version.

Okay thanks! I wonder if we can do that in that PR so that we don't forget it

Reply:
> > It is arbitrary! I will change that in the new version.
> 
> Okay thanks! I wonder if we can do that in that PR so that we don't forget it

I will do so in #2752